File: src/main/java/com/igorgoose/leetcode/java/input/ListNode.java
package com.igorgoose.leetcode.java.input;

public class ListNode {
    public int val;
    public ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

    @Override
    public String toString() {
        return String.format("ListNode(%d)", val);
    }

    public static ListNode fromArray(int[] arr) {
        ListNode head = new ListNode(arr[0]);
        ListNode prev = head;
        for (int i = 1; i < arr.length; i++) {
            ListNode cur = new ListNode(arr[i]);
            prev.next = cur;
            prev = cur;
        }
        return head;
    }

    public static String print(ListNode head) {
        StringBuilder sb = new StringBuilder("[");
        while (head != null) {
            sb.append(head.val).append(", ");
            head = head.next;
        }
        sb.append("]");
        return sb.toString();
    }
}

File: src/main/kotlin/com/igorgoose/leetcode/kotlin/input/ListNode.kt
package com.igorgoose.leetcode.kotlin.input

class ListNode(var `val`: Int) {
    var next: ListNode? = null

    override fun toString() = "ListNode(${`val`})"
}

fun IntArray.toListNodes(): ListNode? = fold((null as ListNode?) to (null as ListNode?)) { (root, last), value ->
    val newNode = ListNode(value)
    if (root == null) {
        return@fold newNode to newNode
    }
    last!!.next = newNode
    root to newNode
}.first

fun ListNode?.listAsString(): String {
    return buildString {
        append(this@listAsString?.toString() ?: "null")
        var cur = this@listAsString?.next
        while (cur != null) {
            append("-> $cur")
            cur = cur.next
        }
    }
}